combainer/serf.go

```diff
@@ -0,0 +1,327 @@
+	if err != nil {
+		return errors.Wrap(err, "bolt store failed")
+	}
+
+	c.raftStore = boltStore
+	c.raftConfig = raft.DefaultConfig()
+	c.raftConfig.NotifyCh = c.leaderCh
+	c.raftConfig.LogOutput = c.log.Logger.Writer()
+	c.raftConfig.StartAsLeader = c.config.StartAsLeader
+	c.updateInterval = interval
+
+	c.log.Info("Create raft")
+	raft, err := raft.NewRaft(c.raftConfig, (*FSM)(c), boltStore, boltStore, snapshots, raftPeers, c.transport)
+	if err != nil {
+		boltStore.Close()
+		c.transport.Close()
+		return errors.Wrap(err, "raft failed")
+	}
+	c.raft = raft
+
+	return nil
+}
+
+// Hosts return names of alive serf members
+func (c *Cluster) Hosts() []string {
+	all := c.serf.Members()
+	alive := make([]string, 0, len(all))
+	for _, m := range all {
+		// that return only alive nodes
+		if m.Status == serf.StatusAlive {
+			alive = append(alive, m.Name)
+		}
+	}
+	return alive
+}
+
+// Members return alive serf members
+func (c *Cluster) Members() []serf.Member {
+	all := c.serf.Members()
+	alive := make([]serf.Member, 0, len(all))
+	for _, m := range all {
+		// that return only alive nodes
+		if m.Status == serf.StatusAlive {
+			alive = append(alive, m)
+		}
+	}
+	return alive
+}
+
+// EventHandler is used to handle events from the serf cluster
+func (c *Cluster) EventHandler() {
+	for {
+		select {
+		case e := <-c.eventCh:
+			switch e.EventType() {
+			case serf.EventMemberJoin:
+				c.nodeJoin(e.(serf.MemberEvent))
+				c.localMemberEvent(e.(serf.MemberEvent))
+			case serf.EventMemberLeave, serf.EventMemberFailed:
+				c.nodeFailed(e.(serf.MemberEvent))
+				c.localMemberEvent(e.(serf.MemberEvent))
+			case serf.EventMemberReap:
+				c.localMemberEvent(e.(serf.MemberEvent))
+			case serf.EventMemberUpdate, serf.EventUser, serf.EventQuery: // Ignore
+			default:
+				c.log.Warnf("unhandled serf event: %#v", e)
+			}
+
+		case <-c.shutdownCh:
+			return
+		}
+	}
+}
+
+// nodeJoin is used to handle join events on the serf cluster
+func (c *Cluster) nodeJoin(me serf.MemberEvent) {
+	for _, m := range me.Members {
+		c.log.WithField("source", "Serf").Infof("Serf join event from %s", m.Name)
+	}
+}
+
+// nodeFailed is used to handle fail events on the serf cluster
+func (c *Cluster) nodeFailed(me serf.MemberEvent) {
+	for _, m := range me.Members {
+		c.log.WithField("source", "Serf").Infof("Serf failed event from %s", m.Name)
+	}
+}
+
+// localMemberEvent is used to reconcile Serf events with the
+// consistent store if we are the current leader.
+func (c *Cluster) localMemberEvent(me serf.MemberEvent) {
+	// Do nothing if we are not the leader
+	if !c.IsLeader() {
+		return
+	}
+
+	// Check if this is a reap event
+	isReap := me.EventType() == serf.EventMemberReap
+
+	// Queue the members for reconciliation
+	for _, m := range me.Members {
+		// Change the status if this is a reap event
+		if isReap {
+			m.Status = statusReap
+		}
+		select {
+		case c.reconcileCh <- m:
+		default:
```

Comment:
If server lost leadership, sending to reconcileCh may block serf EventHandler.
It completely safe because full reconcilation done here https://github.com/combaine/combaine/pull/35/files/b6566b7c146cf672aea6eee87ca71ed5dc723a52#diff-bede81ba08fa838263c52e040287683dR51
```